File: src/save_imu_camera.cpp
/*ros*/
#include <ros/ros.h>
#include <tf/tf.h>
/*msg*/
#include <sensor_msgs/Imu.h>
#include <sensor_msgs/Image.h>
#include <nav_msgs/Odometry.h>
/*C*/
#include <fstream>
/*opencv*/
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

class SaveImageWithIMU{
	private:
		/*node hangle*/
		ros::NodeHandle _nh;
		ros::NodeHandle _nhPrivate;
		/*subscriber*/
		ros::Subscriber _sub_imu;
		ros::Subscriber _sub_image;
		ros::Subscriber _sub_odom;
		/*objects*/
		sensor_msgs::Imu _imu;
		nav_msgs::Odometry _odom_now;
		nav_msgs::Odometry _odom_last;
		std::ofstream _csvfile;
		/*counter*/
		int _counter = 0;
		/*flag*/
		bool _got_first_cb_odom = false;
		/*param*/
		std::string _save_dir_path;
		std::string _save_csv_path;
		std::string _save_img_name;
		int _save_data_limit;
		double _th_diff_position_m;
		double _th_diff_angle_deg;
	public:
		SaveImageWithIMU();
		void callbackIMU(const sensor_msgs::ImuConstPtr& msg);
		void callbackImage(const sensor_msgs::ImageConstPtr& msg);
		void callbackOdom(const nav_msgs::OdometryConstPtr& msg);
		bool hasOdomDiff(nav_msgs::Odometry odom1, nav_msgs::Odometry odom2);
		void publication(void);
		void record(cv_bridge::CvImagePtr cv_ptr);
};

SaveImageWithIMU::SaveImageWithIMU()
	:_nhPrivate("~")
{
	/*param*/
	_nhPrivate.param("save_dir_path", _save_dir_path, std::string("dataset"));
	std::cout << "_save_dir_path = " << _save_dir_path << std::endl;
	_nhPrivate.param("save_csv_path", _save_csv_path, std::string(_save_dir_path + "/save_image_with_imu.csv"));
	std::cout << "_save_csv_path = " << _save_csv_path << std::endl;
	_nhPrivate.param("save_img_name", _save_img_name, std::string("img"));
	std::cout << "_save_img_name = " << _save_img_name << std::endl;
	_nhPrivate.param("save_data_limit", _save_data_limit, 10);
	std::cout << "_save_data_limit = " << _save_data_limit << std::endl;
	_nhPrivate.param("th_diff_position_m", _th_diff_position_m, 10.0);
	std::cout << "_th_diff_position_m = " << _th_diff_position_m << std::endl;
	_nhPrivate.param("th_diff_angle_deg", _th_diff_angle_deg, 30.0);
	std::cout << "_th_diff_angle_deg = " << _th_diff_angle_deg << std::endl;

	/*subscriber*/
	_sub_imu = _nh.subscribe("/imu/data", 1, &SaveImageWithIMU::callbackIMU, this);
	_sub_image = _nh.subscribe("/image_raw", 1, &SaveImageWithIMU::callbackImage, this);
	_sub_odom = _nh.subscribe("/odom", 1, &SaveImageWithIMU::callbackOdom, this);
	/*initialize*/
	_csvfile.open(_save_csv_path, std::ios::app);
	if(!_csvfile){
		std::cout << "Cannot open " << _save_csv_path << std::endl;
		exit(1);
	}
}

void SaveImageWithIMU::callbackIMU(const sensor_msgs::ImuConstPtr& msg)
{
	_imu = *msg;
}

void SaveImageWithIMU::callbackImage(const sensor_msgs::ImageConstPtr& msg)
{
	if(_counter < _save_data_limit){
		if(_got_first_cb_odom && hasOdomDiff(_odom_now, _odom_last)){
			try{
				cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
				record(cv_ptr);
			}
			catch(cv_bridge::Exception& e){
				ROS_ERROR("cv_bridge exception: %s", e.what());
			}
		}
	}
	else if(_csvfile.is_open()){
		_csvfile.close();
		std::cout << _save_csv_path << " was closed" << std::endl;
	}

	/* std::cout << "msg->encoding = " << msg->encoding << std::endl; */
}

void SaveImageWithIMU::callbackOdom(const nav_msgs::OdometryConstPtr& msg)
{
	_odom_now = *msg;
	if(!_got_first_cb_odom){
		_odom_last = *msg;
		_got_first_cb_odom = true;
	}
}

bool SaveImageWithIMU::hasOdomDiff(nav_msgs::Odometry odom1, nav_msgs::Odometry odom2)
{
	/*position*/
	double dx = odom2.pose.pose.position.x - odom1.pose.pose.position.x;
	double dy = odom2.pose.pose.position.y - odom1.pose.pose.position.y;
	double dz = odom2.pose.pose.position.z - odom1.pose.pose.position.z;
	double diff_position_m = sqrt(dx*dx + dy*dy + dz*dz);
	/*rotation*/
	tf::Quaternion q1, q2, q_rel_rot;
	quaternionMsgToTF(odom1.pose.pose.orientation, q1);
	quaternionMsgToTF(odom2.pose.pose.orientation, q2);
	q_rel_rot = q2.inverse()*q1;
	double droll, dpitch, dyaw;
	tf::Matrix3x3(q_rel_rot).getRPY(droll, dpitch, dyaw);
	double diff_angle_deg = q_rel_rot.getAngle()/M_PI*180.0;
	/*print*/
	/* std::cout << "d_xyz : "  */
	/* 	<< dx << ", " */
	/* 	<< dy << ", " */
	/* 	<< dz << std::endl; */
	/* std::cout << "d_rpy : "  */
	/* 	<< droll/M_PI*180.0 << ", " */
	/* 	<< dpitch/M_PI*180.0 << ", " */
	/* 	<< dyaw/M_PI*180.0 << std::endl; */
	/*judge*/
	if(diff_position_m > _th_diff_position_m)	return true;
	if(diff_angle_deg > _th_diff_angle_deg)	return true;
	return false;
}

void SaveImageWithIMU::record(cv_bridge::CvImagePtr cv_ptr)
{
	/*check*/
	std::string save_img_path = _save_dir_path + "/" + _save_img_name + std::to_string(_counter) + ".jpg";
	std::ifstream ifs(save_img_path);
	if(ifs.is_open()){
		std::cout << save_img_path << " already exists" << std::endl;
		exit(1);
	}
	/*save image*/
	cv::imwrite(save_img_path, cv_ptr->image);
	/*record imu*/
	_csvfile 
		<< _imu.linear_acceleration.x << "," 
		<< _imu.linear_acceleration.y << "," 
		<< _imu.linear_acceleration.z << ","
		<< save_img_path << std::endl;
	/*print*/
	double r, p, y;
	tf::Quaternion q;
	quaternionMsgToTF(_odom_now.pose.pose.orientation, q);
	tf::Matrix3x3(q).getRPY(r, p, y);
	std::cout << _counter << " save point: " 
		<< _odom_now.pose.pose.position.x << ", "
		<< _odom_now.pose.pose.position.y << ", "
		<< _odom_now.pose.pose.position.z << ", "
		<< r/M_PI*180.0 << ", " 
		<< p/M_PI*180.0 << ", " 
		<< y/M_PI*180.0 << std::endl;
	std::cout << "imu acc: " 
		<< _imu.linear_acceleration.x << "," 
		<< _imu.linear_acceleration.y << "," 
		<< _imu.linear_acceleration.z << std::endl;
	/*count*/
	++_counter;
	_odom_last = _odom_now;
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "save_image_with_imu");

	SaveImageWithIMU save_image_with_imu;

	ros::spin();
}

File: pysrc/save_imu_camera.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import Imu
from sensor_msgs.msg import Image as ImageMsg
from nav_msgs.msg import Odometry
import tf

from cv_bridge import CvBridge, CvBridgeError

from PIL import Image

import numpy as np
import math
import os
import csv

class SaveImuCameraDepth:
    def __init__(self):
        ## subscriber
        self.sub_imu = rospy.Subscriber("/imu/data", Imu, self.callbackIMU)
        self.sub_imgcolor = rospy.Subscriber("/image_raw", ImageMsg, self.callbackImageColor)
        self.sub_odom = rospy.Subscriber("/odom", Odometry, self.callbackOdom)
        ## msg
        self.imu = Imu()
        self.odom_now = Odometry()
        self.odom_last = Odometry()
        ## cv_bridge
        self.bridge = CvBridge()
        ## img
        self.imgcolor_cv = np.empty(0)
        ## flag
        self.got_first_imu = False
        self.got_first_imgcolor = False
        self.got_first_odom = False
        ## path
        current_dir = os.path.dirname(os.path.abspath(__file__))
        default_rootpath = os.path.join(current_dir, "../dataset/imu_camera/tmp")
        self.rootpath = rospy.get_param("rootpath", default_rootpath)
        self.filename = rospy.get_param("filename", "data_")
        ## counter
        self.counter = 0
        ## threshold
        self.th_diff_position_m = rospy.get_param("th_diff_position_m", 3.0)
        self.th_diff_angle_deg = rospy.get_param("th_diff_angle_deg", 5.0)
        ## print parameter
        print("self.rootpath = ", self.rootpath)
        print("self.filename = ", self.filename)
        print("self.th_diff_position_m = ", self.th_diff_position_m)
        print("self.th_diff_angle_deg = ", self.th_diff_angle_deg)

    def callbackIMU(self, msg):
        self.imu = msg
        if not self.got_first_imu:
            self.got_first_imu = True

    def callbackImageColor(self, msg):
        ## rgb8
        # print(msg.encoding)
        try:
            self.imgcolor_cv = self.bridge.imgmsg_to_cv2(msg, msg.encoding)
            if not self.got_first_imgcolor:
                self.got_first_imgcolor = True
            if self.isReadyToSave():
                self.saveData()
        except CvBridgeError as e:
            print(e)

    def callbackOdom(self, msg):
        self.odom_now = msg
        if not self.got_first_odom:
            self.odom_last = msg
            self.got_first_odom = True

    def isReadyToSave(self):
        if not self.got_first_imu:
            return False
        if not self.got_first_odom:
            return False
        if not self.got_first_imgcolor:
            return False

        if self.hasOdomDiff(self.odom_now, self.odom_last):
            return True
        else:
            return False

    def hasOdomDiff(self, o1, o2):
        ## position
	dx = o2.pose.pose.position.x - o1.pose.pose.position.x
	dy = o2.pose.pose.position.y - o1.pose.pose.position.y
	dz = o2.pose.pose.position.z - o1.pose.pose.position.z
	diff_position_m = math.sqrt(dx*dx + dy*dy + dz*dz)
        ## orientation
        q1 = [
            o1.pose.pose.orientation.x,
            o1.pose.pose.orientation.y,
            o1.pose.pose.orientation.z,
            o1.pose.pose.orientation.w
        ]
        q2_inv = [
            -o2.pose.pose.orientation.x,
            -o2.pose.pose.orientation.y,
            -o2.pose.pose.orientation.z,
            o2.pose.pose.orientation.w
        ]
	q_rel_rot = tf.transformations.quaternion_multiply(q2_inv, q1)
        diff_angle_deg = self.getQuatAngle(q_rel_rot)/math.pi*180.0
        ## judge
        if diff_position_m > self.th_diff_position_m:
            return True
        if diff_angle_deg > self.th_diff_angle_deg:
            return True
        return False

    def saveData(self):
        print("SAVED!")
        print("self.counter = ", self.counter)
        print("self.imgcolor_cv.shape = ", self.imgcolor_cv.shape)
        ## color
        save_imgcolor_name = self.filename + "color_" + str(self.counter) + ".jpg"
        self.saveImageColor(self.imgcolor_cv, save_imgcolor_name)
        ## IMU with images
        self.saveCSV(save_imgcolor_name)
        ## count
        self.counter += 1
        self.odom_last = self.odom_now

    def saveImageColor(self, img_cv, save_name):
        save_path = os.path.join(self.rootpath, save_name)
        if os.path.isfile(save_path):
            print(save_path , " already exists")
            os._exit(1)
        img_pil = Image.fromarray(img_cv)
        img_pil.save(save_path)

    def saveCSV(self, save_imgcolor_name):
        save_csv_path = os.path.join(self.rootpath, "imu_color.csv")
        with open(save_csv_path, "a") as csvfile:
            acc = self.QuatToAcc([
                self.imu.orientation.x,
                self.imu.orientation.y,
                self.imu.orientation.z,
                self.imu.orientation.w
            ])
            writer = csv.writer(csvfile)
            writer.writerow([
                # self.imu.linear_acceleration.x,
                # self.imu.linear_acceleration.y,
                # self.imu.linear_acceleration.z,
                acc[0],
                acc[1],
                acc[2],
                save_imgcolor_name
            ])

    # def cvToPIL(self, img_cv):
    #     img_cv_copy = img_cv.copy()
    #     ## mono
    #     if img_cv_copy.ndim == 2:
    #         pass
    #     ## color
    #     elif img_cv_copy.shape[2] == 3:
    #         img_cv_copy = img_cv_copy[:, :, ::-1]
    #     ## color + alpha
    #     elif img_cv_copy.shape[2] == 4:
    #         img_cv_copy = img_cv_copy[:, :, [2, 1, 0, 3]]
    #     img_pil = Image.fromarray(img_cv_copy)
    #     return img_pil

    def QuatToAcc(self, q):
        g = 1.0
        rpy = tf.transformations.euler_from_quaternion(q)
        acc = (
            -g*math.sin(rpy[1]),
            g*math.sin(rpy[0])*math.cos(rpy[1]),
            g*math.cos(rpy[0])*math.cos(rpy[1])
        )
        print("self.imu.linear_acceleration: ",
            self.imu.linear_acceleration.x, ", ",
            self.imu.linear_acceleration.y, ", ",
            self.imu.linear_acceleration.z
        )
        print("acc = ", acc)
        return acc

    def getQuatAngle(self, q):
        w = q[3].copy()
        w = max(-1.0, min(1.0, w))  #clip
        angle = 2*math.acos(w)
        # angle = math.atan2(math.sin(angle), math.cos(angle))
        # angle = abs(angle)
        return angle


def main():
    rospy.init_node('save_imu_camera', anonymous=True)
    
    save_imu_camera = SaveImuCameraDepth()

    rospy.spin()

if __name__ == '__main__':
    main()
